frame/support/procedural/src/pallet/expand/genesis_build.rs

```diff
@@ -0,0 +1,77 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::pallet::Def;
+use syn::spanned::Spanned;
+
+/// * implement the trait `sp_runtime::BuildModuleGenesisStorage`
+/// * add #[cfg(features = "std")] to GenesisBuild implementation.
+pub fn expand_genesis_build(def: &mut Def) -> proc_macro2::TokenStream {
+	let genesis_config = if let Some(genesis_config) = &def.genesis_config {
+		genesis_config
+	} else {
+		return Default::default()
+	};
+
+	let frame_support = &def.frame_support;
+	let type_impl_gen = &def.type_impl_generics();
+	let type_use_gen = &def.type_use_generics();
+	let trait_use_gen = if def.config.has_instance {
+		quote::quote!(T, I)
+	} else {
+		// `__InherentHiddenInstance` used by construct_runtime here is alias for `()`
+		quote::quote!(T, ())
+	};
+	let gen_cfg_ident = &genesis_config.genesis_config;
+
+	let gen_cfg_use_gen = match (genesis_config.has_trait, genesis_config.has_instance) {
+		(false, false) => (quote::quote!()),
+		(true, false) => (quote::quote!(T)),
+		(true, true) => (quote::quote!(T, I)),
+		(false, true) => unreachable!("Checked by def parser"),
```

Comment:
Why not use an enum to prevent this?

Reply:
What I mean, use a custom enum that doesn't have this variant as this will be rejected by the parser directly.

Reply:
Yes indeed, I introduced a saner type